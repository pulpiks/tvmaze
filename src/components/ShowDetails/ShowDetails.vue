<template>
  <div class="mt-2">
    <!-- eslint-disable-next-line vue/no-v-html -->
    <h1 data-test="show-details-title">{{show.name}}</h1>
    <b-container fluid class="m-0 p-0">
      <b-row no-gutters>
          <b-col cols="12" md="4">
              <b-img :src="show.image.medium" :alt="show.name" fluid
                class="w-100" data-test="show-details-image"></b-img>
          </b-col>
          <b-col cols="12" md="8" class="pl-md-3">
            <h3 class="mb-2 mt-2 mt-md-0">Details</h3>
            <p>
              <strong class="mr-1">Language:</strong>
              <b-badge class="d-inline-block">{{ show.language }}</b-badge>
            </p>
            <p v-if="show.network">
              <strong>Network:</strong> {{show.network.country.name}} / {{show.network.name}}
            </p>
            <p>
              <strong>Schedule:</strong>
              {{ show.schedule.days.join(',') }} at {{show.schedule.time}}
            </p>
            <p>
              <strong class="mr-1">Links:</strong>
              <a :href="show.officialSite" class="mr-2">Official site</a>
              <a :href="show.url" c>Tvmaze link</a>
            </p>
            <p>
              <strong>Rating: </strong>
              <b-icon class="mr-1" icon="star-fill" variant="warning" font-scale="1"></b-icon>
              {{show.rating.average}}
            </p>
            <div class="mb-1">
              <strong class="mr-1">Genres:</strong>
              <div class="d-inline-block">
                <b-badge class="mr-1" v-for="genre in show.genres" :key="genre">{{genre}}</b-badge>
              </div>
            </div>
            <p>
              <strong class="mr-1">Premiere:</strong>{{  formatDate(show.premiered) }}
            </p>
            <p><strong>Status:</strong> {{ show.status }}</p>
            <p v-if="show.summary" v-html="show.summary"></p>
          </b-col>
      </b-row>
    </b-container>

  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import format from 'date-fns/format'
import { TShow } from '@/store/shows'
import { servicesAPI } from '../../api/services'
import { generalErrorHandler } from '../../utils/errorHandler'

@Component({
  name: 'ShowDetails'
})

export default class extends Vue {
  @Prop() private show!: TShow;

  // eslint-disable-next-line class-methods-use-this
  formatDate(date: string | number) {
    try {
      return format(new Date(date), 'dd-MM-yyyy')
    } catch (err) {
      return 'Invalid Date'
    }
  }
}
</script>

<style scoped lang="scss">

</style>
