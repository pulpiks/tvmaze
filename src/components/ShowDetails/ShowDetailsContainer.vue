<template>
  <div class="show-page" data-test="show-page">
    <b-container v-if="show" fluid data-test="show">
      <div v-if="loading" class="d-flex justify-content-center my-2">
        <b-spinner label="Loading..." variant="secondary"></b-spinner>
      </div>
      <div v-else>
        <show-details :show="show"></show-details>
      </div>
    </b-container>
  </div>
</template>

<script lang="ts">
import { servicesAPI } from '@/api/services'
import { generalErrorHandler } from '@/utils/errorHandler'
import { Component, Vue, Prop } from 'vue-property-decorator'
import ShowDetails from './ShowDetails.vue'

@Component({
  name: 'ShowDetailsContainer',
  components: {
    ShowDetails
  }
})

export default class extends Vue {
  private loading = true

  private show = null

  get showId() {
    return this.$route.params.id
  }

  created() {
    this.fetchShow()
  }

  async fetchShow() {
    try {
      this.loading = true
      const { data: show } = await servicesAPI.getShowById(this.showId)
      this.show = show
    } catch (err) {
      generalErrorHandler(this)(err)
    } finally {
      this.loading = false
    }
  }
}
</script>
