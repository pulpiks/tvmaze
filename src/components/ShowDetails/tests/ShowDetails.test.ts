import VueRouter from 'vue-router'
import { BootstrapVue, BootstrapVueIcons } from 'bootstrap-vue'
import {
  createLocalVue, shallowMount, VueClass, Wrapper
} from '@vue/test-utils'
import testData from './testData'

const localVue = createLocalVue()
localVue.use(VueRouter)
localVue.use(BootstrapVue)
localVue.use(BootstrapVueIcons)

// eslint-disable-next-line import/order, import/first
// eslint-disable-next-line import/first
import ShowDetails from '../ShowDetails.vue'

const testAnchors = {
  details: 'show-details',
  name: 'show-details-name',
  image: 'show-details-image'
}


const setUp = (props = {}) => {
  const wrapper = shallowMount(ShowDetails, {
    stubs: ['b-container', 'b-row', 'b-col', 'b-badge', 'b-icon'],
    propsData: props
  })
  return wrapper
}

const findByTestAtrr = (component: any, attr: any) => {
  const wrapper = component.findAll(`[data-test='${attr}']`)
  return wrapper
}

describe('ShowDetails', () => {
  let wrapper: Wrapper<ShowDetails>

  beforeAll(() => {
    wrapper = setUp({
      show: testData
    })
  })

  afterAll(() => {
    wrapper.destroy()
  })

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('render main fields correctly', () => {
    const component = wrapper.find(`[data-test='${testAnchors.details}']`)
    const name = wrapper.find(`[data-test='${testAnchors.name}']`)
    const image = wrapper.find(`[data-test='${testAnchors.image}']`)
    expect(component.exists()).toBe(true)
    expect(name.text()).toEqual(testData.name)
    expect(image.exists()).toBe(true)
  })
})
