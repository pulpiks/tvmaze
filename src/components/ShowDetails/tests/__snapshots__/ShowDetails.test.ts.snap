// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowDetails should match the snapshot 1`] = `
<div
  class="mt-2"
  data-test="show-details"
>
  <h1
    data-test="show-details-name"
  >
    Under the Dome
  </h1>
   
  <b-container-stub
    class="m-0 p-0"
    fluid=""
  >
    <b-row-stub
      no-gutters=""
    >
      <b-col-stub
        cols="12"
        md="4"
      >
        <b-img-stub
          alt="Under the Dome"
          class="w-100"
          data-test="show-details-image"
          fluid=""
          src="http://static.tvmaze.com/uploads/images/medium_portrait/81/202627.jpg"
        />
      </b-col-stub>
       
      <b-col-stub
        class="pl-md-3"
        cols="12"
        md="8"
      >
        <h3
          class="mb-2 mt-2 mt-md-0"
        >
          Details
        </h3>
         
        <p>
          <strong
            class="mr-1"
          >
            Language:
          </strong>
           
          <b-badge-stub
            class="d-inline-block"
          >
            English
          </b-badge-stub>
        </p>
         
        <p>
          <strong>
            Network:
          </strong>
           United States / CBS
          
        </p>
         
        <p>
          <strong>
            Schedule:
          </strong>
          
            Thursday at 22:00
          
        </p>
         
        <p>
          <strong
            class="mr-1"
          >
            Links:
          </strong>
           
          <a
            class="mr-2"
            href="http://www.cbs.com/shows/under-the-dome/"
          >
            Official site
          </a>
           
          <a
            c=""
            href="http://www.tvmaze.com/shows/1/under-the-dome"
          >
            Tvmaze link
          </a>
        </p>
         
        <p>
          <strong>
            Rating: 
          </strong>
           
          <b-icon-stub
            class="mr-1"
            font-scale="1"
            icon="star-fill"
            variant="warning"
          />
          
            6.5
          
        </p>
         
        <div
          class="mb-1"
        >
          <strong
            class="mr-1"
          >
            Genres:
          </strong>
           
          <div
            class="d-inline-block"
          >
            <b-badge-stub
              class="mr-1"
            >
              Drama
            </b-badge-stub>
            <b-badge-stub
              class="mr-1"
            >
              Science-Fiction
            </b-badge-stub>
            <b-badge-stub
              class="mr-1"
            >
              Thriller
            </b-badge-stub>
          </div>
        </div>
         
        <p>
          <strong
            class="mr-1"
          >
            Premiere:
          </strong>
          24-06-2013
          
        </p>
         
        <p>
          <strong>
            Status:
          </strong>
           Ended
        </p>
         
        <p>
          <p>
            <b>
              Under the Dome
            </b>
             is the story of a small town that is suddenly and inexplicably sealed off from the rest of the world by an enormous transparent dome. The town's inhabitants must deal with surviving the post-apocalyptic conditions while searching for answers about the dome, where it came from and if and when it will go away.
          </p>
        </p>
      </b-col-stub>
    </b-row-stub>
  </b-container-stub>
</div>
`;
